cabal-install/src/Distribution/Client/CmdStatus.hs

```diff
@@ -0,0 +1,381 @@
+
+    tsToOriginalTarget ts = tsMap Map.! ts
+
+    mkUnresolvedTarget :: String -> ResolvedTarget
+    mkUnresolvedTarget s = ResolvedTarget s Nothing
+
+-- ----------------------------------------------------------------------------
+-- Target selectors and helpers
+-- ----------------------------------------------------------------------------
+
+-- | This defines what a 'TargetSelector' means for the @status@ command.
+-- It selects the 'AvailableTarget's that the 'TargetSelector' refers to,
+-- or otherwise classifies the problem.
+--
+-- For the @status@ command select all components except non-buildable
+-- and disabled tests\/benchmarks, fail if there are no such
+-- components
+--
+selectPackageTargets :: TargetSelector
+                     -> [AvailableTarget k] -> Either TargetProblem' [k]
+selectPackageTargets targetSelector targets
+
+    -- If there are any buildable targets then we select those
+  | not (null targetsBuildable)
+  = Right targetsBuildable
+
+    -- If there are targets but none are buildable then we report those
+  | not (null targets)
+  = Left (TargetProblemNoneEnabled targetSelector targets')
+
+    -- If there are no targets at all then we report that
+  | otherwise
+  = Left (TargetProblemNoTargets targetSelector)
+  where
+    targets'         = forgetTargetsDetail targets
+    targetsBuildable = selectBuildableTargetsWith
+                         (buildable targetSelector)
+                         targets
+
+    -- When there's a target filter like "pkg:tests" then we do select tests,
+    -- but if it's just a target like "pkg" then we don't build tests unless
+    -- they are requested by default (i.e. by using --enable-tests)
+    buildable (TargetPackage _ _  Nothing) TargetNotRequestedByDefault = False
+    buildable (TargetAllPackages  Nothing) TargetNotRequestedByDefault = False
+    buildable _ _ = True
+
+-- | For a 'TargetComponent' 'TargetSelector', check if the component can be
+-- selected.
+--
+-- For the @build@ command we just need the basic checks on being buildable etc.
+--
+selectComponentTarget :: SubComponentTarget
+                      -> AvailableTarget k -> Either TargetProblem' k
+selectComponentTarget = selectComponentTargetBasic
+
+reportBuildTargetProblems :: Verbosity -> [TargetProblem'] -> IO a
+reportBuildTargetProblems verbosity problems =
+  reportTargetProblems verbosity "status" problems
+
+readTargetSelector :: [PackageSpecifier (SourcePackage (PackageLocation a))]
+                    -> Maybe ComponentKindFilter
+                    -> String
+                    -> IO (Either TargetSelectorProblem TargetSelector)
+readTargetSelector pkgs mfilter targetStr =
+  readTargetSelectors pkgs mfilter [targetStr] >>= \case
+    Left [problem] -> pure $ Left problem
+    Right [ts] -> pure $ Right ts
+    _ -> error $ "CmdStatus.readTargetSelector: invariant broken, more than "
```

Comment:
This should use `fail`

Reply:
Do you expect the exception to be caught?

Reply:
I don't know, but `error` doesn't have predictable semantics.

Reply:
ah, indeed that's a good reason. Is this explained more in-depth somewhere?

Reply:
> ah, indeed that's a good reason. Is this explained more in-depth somewhere?

Not directly, but https://hackage.haskell.org/package/base-4.16.0.0/docs/Control-Exception.html#v:evaluate explains some of it. Basically, `error` can leak through exception handlers (or never trigger at all), depending on where it's called.

E.g. try:

```hs
case Right (error "lol") of
    Right _ -> 1
     _ -> 2
```

And this:

```hs
-- we force the exceptions
(fmap join $ try @SomeException $ fmap (bimap (SomeException . userError) id)
  $ evaluate $ force $ Right @String (error "lol")) >>= \x -> case x of
    Right r -> putStrLn ("not caught " ++ r)
    Left e  -> putStrLn "caught"

-- not forced
(fmap join $ try @SomeException $ fmap (bimap (SomeException . userError) id)
  $ pure $ Right @String (error "lol")) >>= \x -> case x of
    Right r -> putStrLn ("not caught " ++ r)
    Left e  -> putStrLn "caught"